ksqldb-metastore/src/test/java/io/confluent/ksql/metastore/MetaStoreImplTest.java

```diff
@@ -263,6 +265,53 @@ public void shouldDeleteSourceAfterLinkedSourceIsDeleted() {
     assertThat(metaStore.getSource(dataSource1.getName()), is(nullValue()));
   }
 
+  @Test
+  public void shouldDeleteSourceIfAnotherSourceIsLinkedAndRestoreIsInProgress() {
+    // Given:
+    metaStore.putSource(dataSource, false);
+    metaStore.putSource(dataSource1, false);
+    metaStore.addSourceReferences(dataSource1.getName(),
+        Collections.singleton(dataSource.getName()));
+
+    // When:
+    metaStore.deleteSource(dataSource.getName(), true);
+
+    // Then:
+    assertThat(metaStore.getSource(dataSource.getName()), is(nullValue()));
+    assertThat(metaStore.getSource(dataSource1.getName()), is(dataSource1));
+
+    // check that constraints are deleted, but references are not
+    assertThat(metaStore.getSourceConstraints(dataSource.getName()),
+        containsInAnyOrder(ImmutableSet.of().toArray()));
+    assertThat(metaStore.getSourceReferences(dataSource1.getName()),
+        containsInAnyOrder(ImmutableSet.of(dataSource.getName()).toArray()));
+  }
+
+  @Test
+  public void shouldAddReferentialLinkIfRestoreIsInProgress() {
+    // Given:
+    metaStore.putSource(dataSource, false);
+    metaStore.putSource(dataSource1, false);
+    metaStore.addSourceReferences(dataSource1.getName(),
+        Collections.singleton(dataSource.getName()));
+
+    // force source delete if constraint exists
+    metaStore.deleteSource(dataSource.getName(), true);
+
+    // When:
```

Comment:
I'd be tempted to add checks before this to show that `dataSource` has been removed and that the constraint is not present.

Reply:
I added now.